### ISSUE TITLE
Streamable version of voxel grid

### ISSUE THREAD

**Issue by nicolas-chaulet** [PDAL/PDAL#2629]
Hi, I was wondering if there was a way to introduce some sort of streamble version of the voxel grid filter. The use case would be that the input point cloud is 1mm precise say and we apply a voxel grid of 10mm. The output of this would be quite light but with the current setup we have to load all the data in memory before moving forward which seems unecessary to me.

Any idea about how we could go about that?  Is there a way to run part of the steps in a pipeline in streaming mode (the reader in that case) and part of the steps in normal mode? Such a strategy would reduce drastically the memory footprint of those voxel filters because we only need to store in memory one point per voxel unit as opposed to the whole point cloud.

**Comment by chambbj** [comment]
It would be a bit different than the two native voxel-type filters, and more like the "original" PCL filter. I believe it would be possible using online algorithms for computing the mean coordinate per occupied cell. You'd have to scrap (or otherwise interpolate) additional dimension data though. You'd probably also need to either specify your voxel grid offsets or ensure you have trusted bounds in the file header such that you are not depending on computing any of that directly from the data.

**Comment by nicolas-chaulet** [comment]
Regarding the functionality, I think we could ahve the same as the current functionality, just lighter. 
Regarding the size of the grid, you have a point. An option would be to have vaid headers, another way would be to have a dynamic (sparse) grid that would scale with new data points.

**Comment by chambbj** [comment]
Current functionality is to store the single point closest to either the centroid or the voxel center. So, I guess an online update to voxel center is easy enough, as long as the voxel size and position is static. Centroid seems problematic to me, as you won't know the centroid until you've see all of the points, but you don't want to store the points, so how do you get back to the one closest to it? Maybe this is the hybrid step you proposed, but I don't think that will be possible in the current architecture. I don't see how you do any of this with a dynamic grid. Maybe I don't follow what you are proposing though.

**Comment by abellgithub** [comment]
I'm not even sure what you'd be doing with the center algorithm.  How would you mark points as they pass through?  How would you know what points to skip?

**Comment by chambbj** [comment]
> I'm not even sure what you'd be doing with the center algorithm. How would you mark points as they pass through? How would you know what points to skip?

I had imagined each point that is processed first is checked to see if it's corresponding voxel was occupied. If so, and the current point was closer than the existing point, swap them out. If it's not occupied, simply add it.

**Comment by nicolas-chaulet** [comment]
The dynamic grid would be stored in a map with index being the integer coodinate of the voxel. 
- First point comes in, we start with index (0,0,0) and initialise the grid origin as being this first point.
- nth point: we compute the grid index based on the origin of the grid and point spacing. We check wether there is an entry for that voxel, if yes then we update the point representing this voxel based on the chosen algorithm (closest to centroid or centre), if not then we create a new entry in the index
- once all points have been processed we have a point table that contains only the relevent points (closest to centre or to centroid)

Is there something that would not work there? The only issue is that this would still need to process all the points before going to next stage but we only store one point per voxel in memory.

**Comment by abellgithub** [comment]
@nicolas-chaulet : This assumes that you're storing points for later processing.  This isn't how things work.  Points flow through a pipeline in chunks.  When a chunk has been processed, the points are discarded.  There's no way, generally speaking, to hold points for later processing by downstream stages.

**Comment by nicolas-chaulet** [comment]
Yes that's what I suspected. So the question would be how to enable that? Is it a custom Kernel? Or would it require a lot of work to get Pdal to work in streaming mode for part of a pipeline and in no streaming for the rest?

**Comment by nicolas-chaulet** [comment]
Or we embed that into a "super power reader"?

**Comment by abellgithub** [comment]
You can certainly write your own code that does what you want.  You might look at the TileKernel for ideas.

**Comment by nicolas-chaulet** [comment]
Thanks, I'll take a look. What about a reader with downsizing embeded?
Would you be interested in having the contribution?

**Comment by abellgithub** [comment]
When we introduced streaming mode, we eliminated the need for the common filtering demands that might be embedded or attached to a reader.  I think that there are probably more general ways to deal with what you're wanting, but they'd take work.

My approach has been to try to keep things modular and as simple as possible.  I don't know how large your data is, but one obvious solution to your problem is to buy a bunch of memory or run on some cloud server that has sufficient memory.  From a cost perspective, this may be the cheapest way to go depending on your situation.

Another possibility is to use [entwine](https://github.com/connormanning/entwine) for your data and then read it at whatever resolution you desire.

To do something general in software would be significant work.
